dvc/output/__init__.py

```diff
@@ -1,136 +1,390 @@
+from dvc.config import Config
+from dvc.exceptions import DvcException
+from dvc.utils.compat import urlparse, str, builtin_str
+from dvc.remote import Remote
 
-from dvc.output.base import OutputBase
-from dvc.output.s3 import OutputS3
-from dvc.output.gs import OutputGS
-from dvc.output.local import OutputLOCAL
-from dvc.output.hdfs import OutputHDFS
-from dvc.output.ssh import OutputSSH
 
-from dvc.remote import Remote
-from dvc.remote.s3 import RemoteS3
-from dvc.remote.hdfs import RemoteHDFS
-from dvc.remote.local import RemoteLOCAL
-
-OUTS = [
-    OutputHDFS,
-    OutputS3,
-    OutputGS,
-    OutputSSH,
-    # NOTE: OutputLOCAL is the default choice
-]
-
-OUTS_MAP = {
-    Schemes.HDFS: OutputHDFS,
-    Schemes.S3: OutputS3,
-    Schemes.GS: OutputGS,
-    Schemes.SSH: OutputSSH,
-    Schemes.LOCAL: OutputLOCAL,
-}
-
-# NOTE: currently there are only 3 possible checksum names:
-#
-#    1) md5 (LOCAL, SSH, GS);
-#    2) etag (S3);
-#    3) checksum (HDFS);
-#
-# so when a few types of outputs share the same name, we only need
-# specify it once.
-CHECKSUM_SCHEMA = {
-    schema.Optional(RemoteLOCAL.PARAM_CHECKSUM): schema.Or(str, None),
-    schema.Optional(RemoteS3.PARAM_CHECKSUM): schema.Or(str, None),
-    schema.Optional(RemoteHDFS.PARAM_CHECKSUM): schema.Or(str, None),
-}
-
-TAGS_SCHEMA = {schema.Optional(str): CHECKSUM_SCHEMA}
-
-SCHEMA = CHECKSUM_SCHEMA.copy()
-SCHEMA[OutputBase.PARAM_PATH] = str
-SCHEMA[schema.Optional(OutputBase.PARAM_CACHE)] = bool
-SCHEMA[schema.Optional(OutputBase.PARAM_METRIC)] = OutputBase.METRIC_SCHEMA
-SCHEMA[schema.Optional(OutputBase.PARAM_TAGS)] = TAGS_SCHEMA
-SCHEMA[schema.Optional(OutputBase.PARAM_PERSIST)] = bool
-
-
-def _get(stage, p, info, cache, metric, persist=False, tags=None):
-    parsed = urlparse(p)
-
-    if parsed.scheme == "remote":
-        settings = stage.repo.config.get_remote_settings(parsed.netloc)
-        remote = Remote(stage.repo, settings)
-        return OUTS_MAP[remote.scheme](
-            stage,
-            p,
-            info,
-            cache=cache,
-            remote=remote,
-            metric=metric,
-            persist=persist,
-            tags=tags,
+logger = logging.getLogger(__name__)
+
+
+class DoesNotExistError(DvcException):
+    def __init__(self, obj):
+        name = "dependency" if obj.dependency else "output"
+        msg = "{} '{}' does not exist".format(name, obj)
+        super(DoesNotExistError, self).__init__(msg)
+
+
+class IsNotFileOrDirError(DvcException):
+    def __init__(self, obj):
+        name = "dependency" if obj.dependency else "output"
+        msg = "{} '{}' is not a file or directory".format(name, obj)
+        super(IsNotFileOrDirError, self).__init__(msg)
+
+
+class OutputAlreadyTrackedError(DvcException):
+    def __init__(self, path):
+        msg = "output '{}' is already tracked by scm (e.g. git)".format(path)
+        super(OutputAlreadyTrackedError, self).__init__(msg)
+
+
+class Output(object):
```

Comment:
I'm still not sure if I like the idea of joining Dependencies and Outputs (there are lot of `if`-s now: `if self.dependency` which is a sign that it not actually a general enough concept)

Reply:
it should be renamed now, it's confusing to use an output as a dependency. Use `Edge`? since it's a dag.

Reply:
would it be better to create an hierarchy of classes Edge -> (Dependency, Output)?

Reply:
`Edge` is not suitable here, we don't use DAG terms that much. `If`s where there before, with complicated inheritance logic.

Reply:
Output is even less suitable to handle Dependency. Like Output is handling a code dependency. Inheritance was not done properly most likely.

Reply:
Yep, took a second look at it and indeed, I just didn't inherit them properly the last time. Patch incoming.